include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeSubsetSelection2D.h

```diff
@@ -0,0 +1,298 @@
+				double d2 = Intersection(s.end()[-2], s.end()[-1]);
+
+				if (d1 <= d2 || std::abs(d1-d2) < 1.e-10) {//check for numeric stability
+					s.pop_back();
+				} else {
+					break;
+				}
+			}
+			//include the new function and store its index.
+			s.push_back(functions[i]);
+			s.back().index = i;
+			// at the beginning of s are the functions with smallest slope
+			// if the first function in s has a smaller function value for the current 
+			// x_i than the second function,
+			// we can safely remove it as it can not be part of the envelope any more
+			// (We are only looking at function values >=x from now on and thus the larger slope domintes)
+			while (s.size() > 1) {
+				double d1 = s[0].eval(points[i].f1);
+				double d2 = s[1].eval(points[i].f1);
+
+				if (d1 < d2 || std::abs(d1-d2) < 1.e-10) {
+					s.pop_front();
+				} else {
+					break;
+				}
+			}
+			//assign maximum
+			//the functions in s are ordered by function value  
+			// the function with the largest function value is currently at the front
+			h[i] = s[0].eval(points[i].f1);
+			chosen[i] = s[0].index;
+		}
+		return std::make_pair(std::move(h),std::move(chosen));
+	}
+	
+	
+	/// Fast calculation O(n*k) for the hypervolume selection problem
+	/// returns the selected points
+	void hypSSP(std::vector<Point>& front,std::size_t k)const{
+		std::size_t n = front.size();
+		std::vector<LinearFunction> functions(n);
+		
+		std::vector<std::vector<std::size_t> > chosen;
+		std::vector<double>  h(n,0.0);
+		for(std::size_t j=0; j != k-1; ++j) {//compute until k-1 elements are chosen
+			for(std::size_t i=0; i != n; ++i ) {
+				functions[i] = LinearFunction( -front[i].f2, front[i].f1* front[i].f2 + h[i]);
+			}
+			auto result = upperEnvelope(functions, front);
+			h = result.first;
+			chosen.push_back(result.second);
+		}
+		
+		//choose the last element by simply iterating over all elements
+		std::size_t currentIndex = 0;
+		double res = -1;
+		for(std::size_t i=0; i != n; ++i ) {
+			LinearFunction f(-front[i].f2, front[i].f1*front[i].f2 + h[i]);
+			if(f.eval(0)  > res) {
+				res = f.eval(0);
+				currentIndex = i;
+			}
+		}
+		front[currentIndex].selected = true;
+		//iterate backwards to reconstruct chosen indizes
+		for(auto pos = chosen.rbegin(); pos != chosen.rend(); ++pos){
+			currentIndex = (*pos)[currentIndex];
+			front[currentIndex].selected = true;
+		}
+	}
+	
+	template<typename Set>
+	std::vector<Point> createFront(Set const& points, double refX, double refY)const{
+		//copy points using the new reference frame with refPoint at (0,0). also store original index for later
+		std::vector<Point> front;
+		for(std::size_t i = 0; i != points.size(); ++i){
+			front.emplace_back(points[i](0) - refX, points[i](1) - refY,i);
+		}
+		std::sort(front.begin(),front.end());//sort lexicographically
+		//erase dominated points
+		auto newEnd = std::unique(front.begin(),front.end(),[](Point const& x, Point const& y){
```

Comment:
Without the comments this would be entirely crypric, but with the comments in place it is all clear, and actually really neat!